File: integration-tests/pkg/cluster/go/defaults_test.go
package cluster

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/stretchr/testify/require"
)

func TestSchemaDefaults(t *testing.T) {
	val, err := schemaDefault("properties", "kubernetesVersion", "default")
	require.NoError(t, err)
	require.NotNil(t, val)

	val, err = schemaDefault("properties", "machineDefaults", "properties", "hcloud", "properties", "serverType", "default")
	require.NoError(t, err)
	require.NotNil(t, val)

	val, err = schemaDefault("properties", "machineDefaults", "properties", "hcloud", "properties", "datacenter", "default")
	require.NoError(t, err)
	require.NotNil(t, val)

	val, err = schemaDefault("properties", "machines", "items", "properties", "talosImage", "default")
	require.NoError(t, err)
	require.NotNil(t, val)
}

func TestDefaultsAreApplied(t *testing.T) {
	data, err := os.ReadFile(filepath.Join("..", "fixtures", "load-defaults.yaml"))
	require.NoError(t, err)

	raw, err := parseToMap(data)
	require.NoError(t, err)

	err = validateCluster(raw)
	require.NoError(t, err)

	kubeVersion, err := schemaDefault("properties", "kubernetesVersion", "default")
	require.NoError(t, err)
	serverType, err := schemaDefault("properties", "machineDefaults", "properties", "hcloud", "properties", "serverType", "default")
	require.NoError(t, err)
	datacenter, err := schemaDefault("properties", "machineDefaults", "properties", "hcloud", "properties", "datacenter", "default")
	require.NoError(t, err)
	talosImage, err := schemaDefault("properties", "machines", "items", "properties", "talosImage", "default")
	require.NoError(t, err)

	require.Equal(t, kubeVersion, raw["kubernetesVersion"])

	machines, ok := raw["machines"].([]any)
	require.True(t, ok)
	require.Len(t, machines, 1)

	machine := machines[0].(map[string]any)
	hcloud := machine["hcloud"].(map[string]any)
	require.Equal(t, serverType, hcloud["serverType"])
	require.Equal(t, datacenter, hcloud["datacenter"])
	require.Equal(t, talosImage, machine["talosImage"])
}

File: integration-tests/pkg/cluster/python/defaults_test.py
from __future__ import annotations

from cluster.python.defaults import get_default, schema
from cluster.python.validation import validate_cluster
import yaml
from pathlib import Path


def test_schema_defaults_present() -> None:
    assert "properties" in schema()
    assert get_default(["properties", "kubernetesVersion", "default"]) is not None
    assert (
        get_default(["properties", "machines", "items", "properties", "talosImage", "default"])
        is not None
    )
    assert (
        get_default(
            ["properties", "machineDefaults", "properties", "hcloud", "properties", "serverType", "default"]
        )
        is not None
    )
    assert (
        get_default(
            ["properties", "machineDefaults", "properties", "hcloud", "properties", "datacenter", "default"]
        )
        is not None
    )


def test_defaults_are_applied() -> None:
    fixture_path = Path(__file__).resolve().parents[1] / "fixtures" / "load-defaults.yaml"
    with open(fixture_path, "r", encoding="utf-8") as f:
        data = yaml.safe_load(f) or {}

    validate_cluster(data)

    assert data["kubernetesVersion"] == get_default(["properties", "kubernetesVersion", "default"])

    machine = data["machines"][0]
    assert (
        machine["hcloud"]["serverType"]
        == get_default(
            ["properties", "machineDefaults", "properties", "hcloud", "properties", "serverType", "default"]
        )
    )
    assert (
        machine["hcloud"]["datacenter"]
        == get_default(
            ["properties", "machineDefaults", "properties", "hcloud", "properties", "datacenter", "default"]
        )
    )
    assert (
        machine["talosImage"]
        == get_default(["properties", "machines", "items", "properties", "talosImage", "default"])
    )
